controllers/chaosimpl/blockchaos/impl.go

```diff
@@ -17,32 +17,132 @@ package blockchaos
 
 import (
 	"context"
+	"strconv"
+	"time"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
 	"go.uber.org/fx"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
 	impltypes "github.com/chaos-mesh/chaos-mesh/controllers/chaosimpl/types"
 	"github.com/chaos-mesh/chaos-mesh/controllers/chaosimpl/utils"
+	"github.com/chaos-mesh/chaos-mesh/controllers/utils/controller"
+	pb "github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon/pb"
 )
 
 var _ impltypes.ChaosImpl = (*Impl)(nil)
 
 type Impl struct {
 	client.Client
 	Log logr.Logger
+
+	decoder *utils.ContainerRecordDecoder
 }
 
 func (impl *Impl) Apply(ctx context.Context, index int, records []*v1alpha1.Record, obj v1alpha1.InnerObject) (v1alpha1.Phase, error) {
 	impl.Log.Info("blockchaos apply", "record", records[index])
-	// TODO: implement blockchaos apply
+
+	_, _, volumePath, err := controller.ParseNamespacedNameContainerVolumePath(records[index].Id)
+	if err != nil {
+		return v1alpha1.NotInjected, errors.Wrapf(err, "parse container and volumePath %s", records[index].Id)
+	}
+
+	decodedContainer, err := impl.decoder.DecodeContainerRecord(ctx, records[index], obj)
+	pbClient := decodedContainer.PbClient
+	containerId := decodedContainer.ContainerId
+	if pbClient != nil {
+		defer pbClient.Close()
+	}
+	if err != nil {
```

Comment:
Why not check this error as soon as it is returned at L52?

Reply:
This is the tricky part (as all of other container-level injection). It'll need to check whether the `pbClient` should be closed, before returning the error.

I thought there are some problems which makes it hard to clean the `pbClient` inside `DecodeContainerRecord`, but I'm not sure. I'll open an issue to optimize this function (`DecodeContainerRecord`) further. But in this PR, I think we can keep the same with all other implementations.